lib/rules/no-mutation-props.js

```diff
@@ -0,0 +1,227 @@
+const ARRAY_MUTATIONS = ['push', 'pop', 'shift', 'unshift', 'copyWithin', 'fill', 'reverse', 'sort', 'splice'];
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Prevent mutation of this.props',
+      category: 'Possible Errors',
+      recommended: false
+    }
+  },
+
+  create: Components.detect((context, components, utils) => {
+    /**
+     * Gets a unique ID for a node
+     * @param {Object} node The node to get an ID for
+     */
+    const getId = node => node && node.range.join(':');
+
+    const propsNodes = {};
+    /**
+     * Adds a node to the list of nodes that come from this.props
+     * @param {Object} node The node to add
+     */
+    const addToPropsNodes = node => {
+      const id = getId(node);
+
+      if (!id) {
+        return null;
+      }
+
+      if (!propsNodes[id]) {
+        propsNodes[id] = node;
+      }
+      return propsNodes[id];
+    };
+
+    /**
+     * Gets a node if it's been declared as a propsNode
+     * @param {Object} node The node to get
+     */
+    const getPropsNode = node => propsNodes[getId(node)];
+
+    /**
+     * Determines if a given node is from `this.props`
+     * @param {Object} node The node to examine
+     */
+    function isPropsNode (node) {
+      if (!node) {
+        return false;
+      }
+
+      let topObject = node;
+
+      // when looking at an object path, get the top-most object
+      while (topObject.object) {
+        if (topObject.object && topObject.object.object) {
+          topObject = topObject.object;
+        } else {
+          break;
+        }
+      }
+
+      // if this is a simple `this.props` call, we can bail early
+      if (topObject.object && topObject.object.type === 'ThisExpression') {
+        return topObject.property && topObject.property.name === 'props';
+      }
+
+      // if this is a variable, we need to look at the scope to determine if
+      // it is actually a reference to a prop
+      const name = topObject.name || (topObject.object && topObject.object.name);
+      const originalDeclaration = context.getScope().variableScope.set.get(name);
+      if (originalDeclaration) {
+        const originalIdentifiers = originalDeclaration.identifiers;
+        // if the get call returns falsey, we've not set it in our cache
+        // and it can't have have been from props. See `VariableDeclaration`
+        const currentIdentifier = getPropsNode(originalIdentifiers[originalIdentifiers.length - 1]);
+        return currentIdentifier;
+      }
+
+      return false;
+    }
+
+    function setPropMutation(mutation, node) {
+      const component = components.get(utils.getParentComponent());
+      const propMutations = component && component.propMutations
+        ? component.propMutations.concat([mutation])
```

Comment:
is `mutation` an array? If not, the wrapping brackets aren't needed.

Reply:
No, mutation is not an array, it a node.